src/lib.rs

```diff
@@ -384,21 +384,50 @@ impl Drop for Context {
 
 pub struct Socket {
     sock: *mut c_void,
-    closed: bool
+    owned: bool,
 }
 
 unsafe impl Send for Socket {}
 
 impl Drop for Socket {
     fn drop(&mut self) {
-        match self.close_final() {
-            Ok(()) => { debug!("socket dropped") },
-            Err(e) => panic!(e)
+        if self.owned {
+            if unsafe { zmq_sys::zmq_close(self.sock) } == -1i32 {
+                panic!(errno_to_error());
+            } else {
+                debug!("socket dropped");
+            }
         }
     }
 }
 
 impl Socket {
+    /// Consume the Socket and return the raw socket pointer.
+    ///
+    /// Failure to close the raw socket manually or call `from_raw`
+    /// will lead to a memory leak.
+    pub fn into_raw(mut self) -> *mut c_void {
```

Comment:
I have to admit that having just scoured the CZMQ lib, there aren't any instances of CZMQ taking ownership of a Socket. In the past I think it did, but was found to be unnecessary. However if you're doing FFI work, this is a really useful function to have. In another lib I've passed ownership of ZSock instances to C, which could be interchanged for Sockets.

As for `as_mut_ptr`, I think this is equivalent to `borrow_raw`, which is already there. Happy to rename it. I suppose we could remove both `into_raw` and `borrow_raw` in favour of `as_mut_ptr` and use `mem::forget` to avoid freeing the pointer in Rust. I prefer the former methods though as they're more explicit.